File: infrastructures/rest-ddd/csharp/src/Domain/Models/ExampleModel.cs
/**
 * Example Model (BMO) - Business Model Object
 * NO database attributes - pure business logic only.
 */

using System;
using System.Collections.Generic;

namespace Domain.Models
{
    public class ExampleModel : BaseModel
    {
        /**
         * Example domain model demonstrating DDD principles.
         * 
         * Rules:
         * - Member fields prefixed with underscore: _variable or m_variable
         * - NO database/ORM attributes
         * - Contains business logic and validation
         * - Located in /Domain/Models/
         */

        private string? _id;
        private string _name;
        private string _description;
        private ExampleStatus _status;
        private string _ownerId;

        public ExampleModel(
            string p_name,
            string p_description,
            ExampleStatus p_status,
            string p_ownerId,
            string? p_id = null
        )
        {
            _id = p_id;
            _name = p_name;
            _description = p_description;
            _status = p_status;
            _ownerId = p_ownerId;
        }

        // Properties
        public string? Id => _id;
        public string Name => _name;
        public string Description => _description;
        public ExampleStatus Status => _status;
        public string OwnerId => _ownerId;

        // Business logic methods
        public override void Validate()
        {
            if (string.IsNullOrWhiteSpace(_name))
            {
                throw new ArgumentException("Name is required");
            }

            if (_name.Length < 3)
            {
                throw new ArgumentException("Name must be at least 3 characters");
            }

            if (_name.Length > 100)
            {
                throw new ArgumentException("Name must not exceed 100 characters");
            }

            if (string.IsNullOrWhiteSpace(_ownerId))
            {
                throw new ArgumentException("Owner ID is required");
            }
        }

        public void Activate()
        {
            if (_status == ExampleStatus.Active)
            {
                throw new InvalidOperationException("Example is already active");
            }
            _status = ExampleStatus.Active;
        }

        public void Deactivate()
        {
            if (_status == ExampleStatus.Inactive)
            {
                throw new InvalidOperationException("Example is already inactive");
            }
            _status = ExampleStatus.Inactive;
        }

        public void UpdateDescription(string p_description)
        {
            if (p_description.Length > 500)
            {
                throw new ArgumentException("Description must not exceed 500 characters");
            }
            _description = p_description;
        }

        public override Dictionary<string, object> ToDictionary()
        {
            return new Dictionary<string, object>
            {
                { "id", _id ?? string.Empty },
                { "name", _name },
                { "description", _description },
                { "status", _status.ToString() },
                { "ownerId", _ownerId }
            };
        }
    }

    public enum ExampleStatus
    {
        Active,
        Inactive,
        Pending
    }
}

File: infrastructures/rest-ddd/typescript/src/domain/models/ExampleModel.ts
/**
 * Example Model (BMO) - Business Model Object
 * NO database attributes - pure business logic only.
 */

import { BaseModel } from './BaseModel';

export class ExampleModel extends BaseModel {
  /**
   * Example domain model demonstrating DDD principles.
   * 
   * Rules:
   * - Member fields prefixed with underscore: _variable
   * - NO database/ORM attributes
   * - Contains business logic and validation
   * - Located in /domain/models/
   */

  private _id?: string;
  private _name: string;
  private _description: string;
  private _status: ExampleStatus;
  private _ownerId: string;

  constructor(
    p_name: string,
    p_description: string,
    p_status: ExampleStatus,
    p_ownerId: string,
    p_id?: string
  ) {
    super();
    this._id = p_id;
    this._name = p_name;
    this._description = p_description;
    this._status = p_status;
    this._ownerId = p_ownerId;
  }

  // Getters
  get id(): string | undefined {
    return this._id;
  }

  get name(): string {
    return this._name;
  }

  get description(): string {
    return this._description;
  }

  get status(): ExampleStatus {
    return this._status;
  }

  get ownerId(): string {
    return this._ownerId;
  }

  // Business logic methods
  validate(): void {
    if (!this._name || this._name.trim().length === 0) {
      throw new Error('Name is required');
    }

    if (this._name.length < 3) {
      throw new Error('Name must be at least 3 characters');
    }

    if (this._name.length > 100) {
      throw new Error('Name must not exceed 100 characters');
    }

    if (!this._ownerId) {
      throw new Error('Owner ID is required');
    }
  }

  activate(): void {
    if (this._status === ExampleStatus.ACTIVE) {
      throw new Error('Example is already active');
    }
    this._status = ExampleStatus.ACTIVE;
  }

  deactivate(): void {
    if (this._status === ExampleStatus.INACTIVE) {
      throw new Error('Example is already inactive');
    }
    this._status = ExampleStatus.INACTIVE;
  }

  updateDescription(p_description: string): void {
    if (p_description.length > 500) {
      throw new Error('Description must not exceed 500 characters');
    }
    this._description = p_description;
  }

  toObject(): Record<string, any> {
    return {
      id: this._id,
      name: this._name,
      description: this._description,
      status: this._status,
      ownerId: this._ownerId
    };
  }
}

export enum ExampleStatus {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  PENDING = 'PENDING'
}
